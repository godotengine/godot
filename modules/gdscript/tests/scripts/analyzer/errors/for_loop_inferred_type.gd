class Iterator:
	func _iter_init(_count):
		return true
	func _iter_next(_count):
		return false
	func _iter_get(_count) -> StringName:
		return &"custom"

enum { ENUM_VALUE = 1 }

func test():
	# Literals.

	for x in true:
		pass

	for x in 1:
		if x is String:
			pass

	# Constants.

	const CONSTANT_INT = 1
	for x in CONSTANT_INT:
		if x is String:
			pass

	const CONSTANT_FLOAT = 1.0
	for x in CONSTANT_FLOAT:
		if x is String:
			pass

	# Hard types.

	var hard_int := 1
	for x in hard_int:
		if x is String:
			pass

	var hard_float := 1.0
	for x in hard_float:
		if x is String:
			pass

	var hard_string := "a"
	for x in hard_string:
		if x is int:
			pass

	# Other.

	for x in ENUM_VALUE:
		if x is String:
			pass

	var hard_iterator := Iterator.new()
	for x in hard_iterator:
		if x is int:
			pass
