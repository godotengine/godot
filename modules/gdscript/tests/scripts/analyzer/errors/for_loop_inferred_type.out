GDTEST_ANALYZER_ERROR
>> ERROR at line 14: Unable to iterate on value of type "bool".
>> ERROR at line 18: Expression is of type "int" so it can't be of type "String".
>> ERROR at line 25: Expression is of type "int" so it can't be of type "String".
>> ERROR at line 30: Expression is of type "float" so it can't be of type "String".
>> ERROR at line 37: Expression is of type "int" so it can't be of type "String".
>> ERROR at line 42: Expression is of type "float" so it can't be of type "String".
>> ERROR at line 47: Expression is of type "String" so it can't be of type "int".
>> ERROR at line 53: Expression is of type "int" so it can't be of type "String".
>> ERROR at line 58: Expression is of type "StringName" so it can't be of type "int".
